internal/cmdrender/executor.go

```diff
@@ -66,6 +67,13 @@ func (e *Executor) Execute(ctx context.Context) error {
 	}
 
 	pkgWriter := &kio.LocalPackageReadWriter{PackagePath: string(root.pkg.UniquePath)}
+
+	// format resources before writing
+	_, err = filters.FormatFilter{UseSchema: true}.Filter(resources)
```

Comment:
So it seems like this means we will reorder fields as part of running the render command. Have we made a clear decision that we want to force a single ordering for all resources in kpt? If so, can we add a clear description of the rules we reorder based on?

Reply:
Described in the PR description. What should I elaborate on further ? And do you mean update PR description or add comment with full description ?

Reply:
So I think my question/concern is around three areas:
 * How does this work with CRs where we might not have an openapi schema? Until we have a good solution for finding CRDs/schemas, this is probably a situation many users will run into. Based on the description, seems like we would fall back to lexicographical ordering in this case. Although looking at the tests, seems like we would recognize the metadata section for CRs, so those fields will be ordered based on openapi definition (tests should `labels` being placed after `name`).
 * Does the resulting order mostly match the ordering that users expect from k8s resources? I don't know how much thought went into the ordering of fields in the openapi schema for k8s resources. I also expect there to some exceptions to using the ordering based on the openapi schema, mostly around associative maps (although it might be that the merge-key is the first property in most cases).
 * How do we handle this for the full workflow? It seems like this means we should encourage users to format their resources based on this ordering before publishing to avoid a poor diff experience. Do we still plan on having a separate `fmt` command?

If the formatted resources mostly align with what users want, I agree that enforcing it solves many challenges.

Reply:
> So I think my question/concern is around three areas:
> 
> * How does this work with CRs where we might not have an openapi schema? Until we have a good solution for finding CRDs/schemas, this is probably a situation many users will run into. Based on the description, seems like we would fall back to lexicographical ordering in this case. Although looking at the tests, seems like we would recognize the metadata section for CRs, so those fields will be ordered based on openapi definition (tests should `labels` being placed after `name`).

Yes. For unknown CR's the fields after the metadata section will be sorted in lexicographical order. 

> * Does the resulting order mostly match the ordering that users expect from k8s resources? I don't know how much thought went into the ordering of fields in the openapi schema for k8s resources. I also expect there to some exceptions to using the ordering based on the openapi schema, mostly around associative maps (although it might be that the merge-key is the first property in most cases).

There might be some exception but k8s resources schema is the closest we can get.

> * How do we handle this for the full workflow? It seems like this means we should encourage users to format their resources based on this ordering before publishing to avoid a poor diff experience. Do we still plan on having a separate `fmt` command?
> 
Yes. Publishers should format resources before publishing to avoid unwanted diffs. The command will be deprecated but we will be providing a [function](https://github.com/GoogleContainerTools/kpt-functions-catalog/pull/242)(essentially same thing).

> If the formatted resources mostly align with what users want, I agree that enforcing it solves many challenges.